File: packages/tools/src/tools/write.ts
import { writeFile, mkdir } from "node:fs/promises";
import { dirname } from "node:path";
import type { ToolDefinition, ToolResult } from "@forge/types";

export const writeTool: ToolDefinition = {
  name: "Write",
  description:
    "Writes content to a file at the specified path. Creates parent directories " +
    "if they don't exist. Overwrites the file if it already exists.",
  inputSchema: {
    type: "object",
    properties: {
      file_path: {
        type: "string",
        description: "The absolute path to the file to write",
      },
      content: {
        type: "string",
        description: "The content to write to the file",
      },
    },
    required: ["file_path", "content"],
  },
  annotations: { readOnly: false },
  handler: async (input): Promise<ToolResult> => {
    const filePath = input.file_path as string;
    const content = input.content as string;

    try {
      // Create parent directories if they don't exist
      const dir = dirname(filePath);
      await mkdir(dir, { recursive: true });

      // Write file
      await writeFile(filePath, content, "utf-8");

      const byteCount = Buffer.byteLength(content, "utf-8");
      return {
        content: [
          {
            type: "text",
            text: `Successfully wrote ${byteCount} bytes to ${filePath}`,
          },
        ],
      };
    } catch (err) {
      const error = err as Error;
      return {
        content: [
          {
            type: "text",
            text: `Failed to write file: ${error.message}`,
          },
        ],
        isError: true,
      };
    }
  },
};

File: internal/tools/write.go
package tools

import (
	"fmt"
	"os"
	"path/filepath"

	"github.com/jelmersnoeck/forge/internal/types"
)

// WriteTool returns the Write tool definition.
func WriteTool() types.ToolDefinition {
	return types.ToolDefinition{
		Name:        "Write",
		Description: "Writes content to a file, creating parent directories if needed. Overwrites existing files.",
		InputSchema: map[string]any{
			"type": "object",
			"properties": map[string]any{
				"file_path": map[string]any{
					"type":        "string",
					"description": "Absolute path to the file",
				},
				"content": map[string]any{
					"type":        "string",
					"description": "Content to write",
				},
			},
			"required": []string{"file_path", "content"},
		},
		Handler:     writeHandler,
		ReadOnly:    false,
		Destructive: false,
	}
}

func writeHandler(input map[string]any, ctx types.ToolContext) (types.ToolResult, error) {
	filePath, err := requireString(input, "file_path")
	if err != nil {
		return types.ToolResult{IsError: true}, err
	}
	content, ok := input["content"].(string)
	if !ok {
		return types.ToolResult{IsError: true}, fmt.Errorf("content is required")
	}

	if isEnvFile(filePath) {
		return envFileError(filePath), nil
	}

	dir := filepath.Dir(filePath)
	if err := os.MkdirAll(dir, 0755); err != nil {
		return errResultf("failed to create parent directories: %v", err)
	}

	if err := os.WriteFile(filePath, []byte(content), 0644); err != nil {
		return errResultf("failed to write file: %v", err)
	}

	if ctx.ReadState != nil {
		delete(ctx.ReadState, filePath)
	}

	return textResult(fmt.Sprintf("wrote %d bytes to %s", len(content), filePath)), nil
}
